runner/src/main/kotlin/org/jetbrains/idea/inspections/IdeaRunner.kt

```diff
@@ -0,0 +1,272 @@
+        // NB: IdeaApplication.getStarter() from IJ community contains the same call
+        val enabledPlugins = PluginManagerCore.getPlugins().map { it.name to it.pluginId }.toMap()
+        val disabledPlugins = PluginManagerCore.getDisabledPlugins().toList()
+        logger.info("InspectionPlugin: Enabled plugins: $enabledPlugins")
+        logger.info("InspectionPlugin: Disabled plugins $disabledPlugins")
+        return ApplicationManagerEx.getApplicationEx().apply { load() }
+    }
+
+    private fun generateSystemPath(buildNumber: String, usesUltimate: Boolean): String {
+        val homeDir = System.getProperty(USER_HOME).replace("\\", "/")
+        val buildPrefix = (if (usesUltimate) "U_" else "") + buildNumber.replace(".", "_")
+        var path: String
+        var code = 0
+        do {
+            code++
+            path = "$homeDir/.IntellijIDEAInspections/${buildPrefix}_code$code/system"
+            val file = File(path)
+            if (!file.exists()) {
+                file.mkdirs()
+            }
+            val systemMarkerFile = File(file, SYSTEM_MARKER_FILE)
+            systemMarkerFile.createNewFile()
+            // To prevent usages by multiple processes
+            val lock = try {
+                systemPathMarker = FileChannel.open(systemMarkerFile.toPath(), StandardOpenOption.WRITE)
+                systemPathMarker?.tryLock()
+            } catch (e: IOException) {
+                logger.warn("InspectionPlugin: IO exception while locking: ${e.message}")
+                throw RunnerException("EXCEPTION caught in inspection plugin (IDEA system dir lock): $e", e)
+            }
+            if (lock == null && code == 256) {
+                throw RunnerException("Cannot create IDEA system directory (all locked)")
+            }
+        } while (lock == null)
+        return path
+    }
+
+    private fun ideaAcquireIfNeeded() {
+        val tempDir = File(System.getProperty("java.io.tmpdir"))
+        val pluginDir = File(tempDir, "inspection-plugin")
+        val ideaLockDir = File(pluginDir, "ideaLock")
+        if (!ideaLockDir.exists()) ideaLockDir.mkdirs()
+        val bean = ManagementFactory.getRuntimeMXBean()
+        val lockName = "${bean.name}.lock"
+        ideaLockDir.listFiles()?.firstOrNull()?.let { lockFile ->
+            ideaLockChannel = FileChannel.open(lockFile.toPath(), StandardOpenOption.WRITE)
+            ideaLock = try {
+                ideaLockChannel?.tryLock()
+            } catch (ignore: OverlappingFileLockException) {
+                null
+            }
+            when (ideaLock) {
+                null -> logger.info("InspectionPlugin: IDEA lock ${lockFile.name} acquire skipped")
+                else -> logger.info("InspectionPlugin: IDEA lock ${lockFile.name} acquired")
+            }
+            if (lockFile.name == lockName) return
+            if (ideaLock == null) throw RunnerException("IDEA already used in other process.")
```

Comment:
Different processes should be able to use the same version of IDEA (by creating different instances of it with different system paths).